### ISSUE TITLE
Song cache should not cache partial songs

### ISSUE THREAD

**Issue by gechoto** [OuterTune/OuterTune#133]
### Checklist

- [X] I've checked that there is no other issue about this feature request.
- [X] This issue contains only one feature request.
- [X] The title of this issue accurately describes the feature request.

### Feature description

Hi, the song cache from InnerTune is missing. But this was a really nice feature.

### Why do you want this feature?

- Save data
- Make less requests to YT (maybe improves getting rate limited and getting blocked without login)

### Additional information

I guess you removed it because partially cached songs do not work. InnerTune has the same issue:
https://github.com/z-huang/InnerTune/issues/1726

However I think completely removing the cache is a bit extreme.
I have this idea of keeping the cache like InnerTune - but only use it for fully cached songs.

So the logic could be like this:
- if a song is fully (100%) in cache -> play it from cache
- if a song is only partially in cache (< 100%) -> remove it from cache and re-request it

This should work and be a nice compromise which helps many users.

**Comment by gechoto** [comment]
When programming the song cache please remember to also remove songs from the `format` table in the DB when they are removed from cache.
Otherwise we run into this problem:
https://github.com/DD3Boh/OuterTune/issues/134

**Comment by mikooomich** [comment]
Cache was removed mainly because the workaround to avoid caching local songs (somehow way more difficult than it seems) doesn't play well with YouTube and caused playback errors. 

See here https://github.com/DD3Boh/OuterTune/issues/16

I'll see if it's feasible to bring it back

**Comment by lucasmz-dev** [comment]
Yeah I definitely noticed this (songs taking a while to load even though I just listened to them before), and given it's more common for me to be using OuterTune on-the-go rather than when I'm on Wi-Fi, this was pretty important, it was a very nice feature in InnerTune that I end up really missing because I don't really wanna go around downloading what I *think* I'm listening to, I just want it to be done automatically 🙃

Especially with how OuterTune integrates with YouTube Music itself, I seem to be able to rely on YouTube's algorithm for what I listen to.
